### DISCUSSION TITLE
nushell does not properly recognize the environment variable Path

### DISCUSSION THREAD

**Discussion by baojd42** [nushell/nushell#13427] [Q&A] [answered]
I use winget to install pnpm. After the installation, I can see the bin path is written into the environment variable Path. However, when I do `echo $env.Path` in nushell, I can not see pnpm'b bin path. And executing pnpm gives error "executable was not found". What could be the reason? Thanks for any help.

**Comment by baojd42** [comment]
Restart the computer, and the path is there...

One thing to mention, I restarted the terminal several times.

**Reply by Lockszmith-GH** [reply]
see [my answer](https://github.com/nushell/nushell/discussions/13427#discussioncomment-10129396) as to why.

**Comment by ghost** [comment]
Im not a windows user, but a theory as to what might have gone wrong here:

winget might not (fully) support nushell and just wrote the PATH addition to your bash/.. config and not your nushell config.  
with this restarting nushell would (usually) not fix it.  
a lot of env variables (including PATH) get set on a system or user level on boot/login and get used as default by other programs (such as nu).  
you could check if the nupm path is in your `$nu.env-path` or `$nu.config-path` and if its not this is probably what happened.

another potentially issue might be that nushell didnt restart due to a auto-attaching TMUX or something.

**Reply by Lockszmith-GH** [reply]
Windows is 'a bit' different. See [my answer](https://github.com/nushell/nushell/discussions/13427#discussioncomment-10129396).

**Comment by NotTheDr01ds** [comment]
> I restarted the terminal several times.

Windows Terminal itself or the Nushell tab/session?

I'm guessing there may have been something leftover from Windows Terminal since it should read the environment on restart.

**Reply by baojd42** [reply]
The terminal. I closed and reopened it.

**Reply by Lockszmith-GH** [reply]
See my [answer below](https://github.com/nushell/nushell/discussions/13427#discussioncomment-10129396). Restarting terminal isn't enough, as the envrionment persissts on the `explorer.exe` level.

**Comment by Lockszmith-GH** [comment, marked as answer]
A Windows user here.

The way the environment works in Windows (I believe this goes back to Windows XP, maybe even earlier NT versions) in a way, is very similar to how it works in Linux.

Each process (by default) shares the environment of the process that initiated it.
When the system loads there are two sets of registry values each enumerating the envrionment vairables and their values. One list of registry values for the System variables, and the other for the User specific variables.

The only "official" way to relead the system envrionment is to Logoff and then Login again.

This is because the top most process that shares it's envrionment is `explorer.exe` which provides both the default UI shell (The Taskbar, Start Menu, System Tray and the Desktop) as well as Windows Explorer - which is the interface you use to interact with folders.

The quick-and-dirty way to reload the environment without a logoff is to kill the `explorer.exe` process with Task Manager.
This will cause your taskbar and Desktop to dissapear. By default windows launches `explorere.exe` automatically when it dies, but if not, from Task Manager, use File -> Run to run `explorere.exe` again.
When `explorer.exe` loads again, it reads the envrionment from the registry, and anything it will invoke (like a new terminal running a new `nu` session) will inherit the fresh envrionment.

There is a GitHub project [badrelmers/RefrEnv](/badrelmers/RefrEnv) that has shell scripts to fix this (bash, zsh, cmd and powershell).
That is probably the cleanset way to resolve this in Windows today. It might need a nu-shell contribution.

Hope this helps.

**Reply by fdncred** [reply]
Right. The Windows registry was introduced in Windows 3.1. I remember when it came out, lol. We could probably do something like the chocolatey script by using our `load-env` command and our `registry query` command.

As indicated in the help for `registry query` you can see the HKLM and HKCU environment settings with these commands.
```nushell
registry query --hklm 'SYSTEM\CurrentControlSet\Control\Session Manager\Environment'
registry query --hkcu environment
```
You'd probably need to set those to a variable, combine each item and uniq it, only get the name and value, transpose -ird and try to do a load-env. I'm not sure it'll work but it's possible that it could. Some time and fiddling is needed.

**Reply by fdncred** [reply]
I threw this together real quick.
```nushell
# User environment variables....take precedence over system environment variables.
# except
# The Path is constructed from the system path (hklm) and the user path (hkcu). The
# User path is appended to the system path.
let hklm_env = registry query --hklm 'SYSTEM\CurrentControlSet\Control\Session Manager\Environment'
let hkcu_env = registry query --hkcu environment
let hklm_path = $hklm_env | where name == Path | get value.0
let hkcu_path = $hkcu_env | where name == Path | get value.0
let global_env = ($hklm_env | where name != Path) ++ ($hkcu_env | where name != Path)
let global_path = $'($hklm_path)(char esep)($hkcu_path)' | split row (char esep) | uniq
let new_env = $global_env | reject type | transpose -ird | merge {Path: $global_path}
load-env $new_env
```
